### PR TITLE
[NFC] Poison references at -Onone after potentially shortening object lifetime

### PR DESCRIPTION
**Pull request by atrick**
This PR adds support for poison debug values to the -mandatory-copy-propagation pass. The pass isn't enabled in this PR, so it's intended to be NFC.

---

To eliminate ARC traffic, the optimizer reorders object destruction. This changes observable program behavior. If a custom deinitializer produces side effects, code may observe those side effects earlier after optimization. Similarly, code that dereferences a weak reference may observe a 'nil' reference after optimization, while the unoptimized code observed a valid object.

Developers have overwhelmingly requested that object lifetimes have similar behavior in -Onone and -O builds in order to find and diagnose program bugs involving weak references and other lifetime assumptions.

Enabling the copy propagation at -Onone is simply a matter of flipping a switch. -Onone runtime and code size will improve. By design, copy propagation, has no direct affect on compile time. It will indirectly improve optimized compile times, but in debug builds, it simply isn't a factor.

To support debugging, a "poison" flag is added to new destroy_value instructions generated by copy propagation.
When OwnershipModelEliminator lowers destroy_value [poison] it will generate new debug_value instructions with a “poison” flag:

```
debug_value %ref, let, name "a"
debug_value %ref, let, name "b"
destroy_value [poison] %ref
```
--->
```
debug_value %ref, let, name "a"
debug_value %ref, let, name "b"
strong_release %ref
debug_value [poison] %ref, let, name "a"
debug_value [poison] %ref, let, name "b"
```

At -Onone, IRGen generates a shadow stack variable for each debug variable. At the point in the program where a poison debug_value occurs, IRGen will recursively overwrites all references within the shadow stack variable to a sentinel pointer value that flags that the shadow variable’s underlying variable is no longer live. This value should be a valid pointer according to ABI rules to preserve spare bits but should be guaranteed to be a protected page. This will handle all value types with a visible layout.

### CODE REVIEW

**Comment by atrick** [comment]
@aschwaighofer IRGen changes are all in this commit https://github.com/apple/swift/pull/36389/commits/2b495061dd0870e4c95b8a9f7f89d8f6f456e566

**Comment by atrick** [comment]
@adrian-prantl these test cases could be reused for lldb: https://github.com/apple/swift/pull/36389/commits/c4d341b449fafb0d5f1f41d18069621666dd1fbd

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci benchmark

**Comment by atrick** [comment]
@swift-ci test source compatibility

**Comment by atrick** [comment]
I converted this to NFC so I can merge in earlier and avoid later conflicts or CI issues related to the new test cases.

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci smoke test macOS

**Review by eeckstein — COMMENTED** [review]
Just a few minor things.
I didn't look at the IRGen changes

**Comment by atrick** [comment]
@swift-ci smoke test macOS

**Comment by atrick** [comment]
@swift-ci smoke test macOS

**Comment by atrick** [comment]
@swift-ci clean smoke test macOS

File: lib/SILOptimizer/Utils/CanonicalOSSALifetime.cpp [hunk 1]
```diff
@@ -602,15 +604,23 @@ void CanonicalizeOSSALifetime::extendLivenessThroughOverlappingAccess() {
...
+  SILBasicBlock *predBB, SILBasicBlock *succBB, bool needsPoison) {
+
   assert(succBB->getSinglePredecessorBlock() == predBB
          && "value is live-out on another predBB successor: critical edge?");
 
-  SILBuilderWithScope builder(succBB->begin());
-  auto loc = RegularLocation::getAutoGeneratedLocation(succBB->begin()->getLoc());
-  auto *di = builder.createDestroyValue(loc, def);
+  auto pos = succBB->begin();
+  while (pos->isDebugInstruction()) {
```
[the remark this answers is not in the document]
Comment (atrick) [reply lib/SILOptimizer/Utils/CanonicalOSSALifetime.cpp:614]: I hacked that in thinking that lldb user could break at the beginning of an if/else block and still inspect variables from the outer scope. Then I forgot to test it. Turns out setting a breakpoint at the begging of the block doesn't work anyway. For now, I'm removing the code to advance past debug_values. In the future, we could consider pushing `destroy_value [poison]` down further into block where it isn't used, but I can't come up for a good use case now.

File: lib/SILOptimizer/Utils/CanonicalOSSALifetime.cpp [hunk 2]
```diff
@@ -867,10 +955,61 @@ void CanonicalizeOSSALifetime::rewriteCopies() {
   }
 }
 
+/// Insert poison destroys after each consume that needs poison.
+void CanonicalizeOSSALifetime::injectPoison() {
+  for (SILInstruction *consume : consumes.getNeedsPoisonConsumes()) {
+    auto createDestroy = [&](SILBuilder &builder) {
+      auto pos = builder.getInsertionPoint();
+      while (pos->isDebugInstruction()) {
```
[the remark this answers is not in the document]
Comment (atrick) [reply lib/SILOptimizer/Utils/CanonicalOSSALifetime.cpp:963]: I removed this code to sink poison past debug instructions. I had added it in a fit of paranoia, but I can't find an example where it actually helps.

**Comment by atrick** [comment]
> Just a few minor things.
> I didn't look at the IRGen changes

Thanks for looking. I'm not really happy with this code anymore--it wasn't supposed to grow all these extra features. I'd like "find final consumes" to be a stand-alone algorithm without any extra book-keeping and heuristics. I need to think about how to do that efficiently and give it a nice interface, and for now there are other pieces of code in much more need of simplification. I do have a rewrite of ValueLifetimeAnalysis, which is quite similar, waiting to checkin.

**Comment by atrick** [comment]
@swift-ci smoke test

**Comment by atrick** [comment]
@swift-ci smoke test

**Pull request merged**